### ISSUE TITLE
Sync all Taskwarrior tasks with Google Tasks

### ISSUE THREAD

**Issue by bergercookie** [bergercookie/syncall#103]
### Discussed in https://github.com/bergercookie/syncall/discussions/102

<div type='discussions-op-text'>

<sup>Originally posted by **shymega** January  5, 2024</sup>
HI!

Great project (albeit I have to use it from Git, due to the Cython3 PyYAML dep) so far.

I'm having trouble syncing all of my Taskwarrior tasks - I wanted to select everything to sync to Google Tasks.

Is that possible?

Thanks :)</div>

**Comment by bergercookie** [comment]
@shymega please have a look at the latest state of the master branch. it should resolve both the PyYAML installation exception and, for tw_gtasks_sync it should add a few different flags to the executable one of which is the `--all`

```
 --all, --taskwarrior-all-tasks  Sync all taskwarrior tasks [potentially very
                                 slow]
```

Let me know if that doesn't work for you. It will make its way into an actual PyPI release in the next few weeks.

**Comment by shymega** [comment]
Hey @bergercookie - I just cloned, and I can't see any commits that add `--all`. Not sure if you pushed, or if I'm being stupid.

**Comment by bergercookie** [comment]
Sorry, typed this too fast and haven't merged to master yet. Have a look at branch 59-virtual-tags instead :)

**Comment by shymega** [comment]
After I install with `pipx install --force ".[google,tw]"`, I get the following error:

```shell
syncall on  59-virtual-tags is 📦 v0.0.0 via 🐍 v3.11.7 on ☁️  (eu-west-2) took 10s
❯ tw_gtasks_sync --help
2024-01-20 22:03:22.769 | ERROR    | syncall.app_utils:inform_about_app_extras:235 -
You have to install the google,tw extras for tw_gtasks_sync to work.
With pip, you can do it with something like: "pip3 install syncall[google,tw]"
Exiting.
```

I'm on NixOS, so I have to use virtualenvs until I get around to packaging `syncall`. But this did work on `master`, so I suspect a regression.

**Comment by bergercookie** [comment]
Probably a dependency issue
What do the following lines say?

```
python3 -c 'from syncall import GTasksSide, TaskWarriorSide'
python3 -c 'from syncall.taskwarrior.taskwarrior_side import TaskWarriorSide'
python3 -c 'from syncall.google.gtasks_side import GTasksSide'
```

**Comment by bergercookie** [comment]
This should now be supported across all the apps in this repo that interact with Taskwarrior. Use the following flag to specify that you want all your taskwarrior tasks to be synchronized.

```
 --all, --taskwarrior-all-tasks  Sync all taskwarrior tasks (potentially very slow).
```

@shymega let me know if this still doesn't work and I can help debug this.

**Comment by shymega** [comment]
> This should now be supported across all the apps in this repo that interact with Taskwarrior. Use the following flag to specify that you want all your taskwarrior tasks to be synchronized.
> 
> ```
>  --all, --taskwarrior-all-tasks  Sync all taskwarrior tasks (potentially very slow).
> ```
> 
> This feature should work on latest master - commit [0893f12](https://github.com/bergercookie/syncall/commit/0893f12870aff5242118993c70f61bbfc85544cb) - and should make its way to a pypi release in the coming week(s)
> 
> @shymega let me know if this still doesn't work and I can help debug this.

Thank you for implementing my feature request! Appreciated. Just replied with the output for troubleshooting you requested, too.

Cheers.

**Comment by bergercookie** [comment]
> ModuleNotFoundError: No module named 'packaging'

Yeah, that's a missing dependency. You can `pip install packaging` and that should do it.
That dependency comes from the newly added [taskw-ng](https://github.com/bergercookie/taskw-ng) package. I'll bump its version and it should go away

**Comment by shymega** [comment]
Thanks! I'm just having to create a new OAuth app, as the bundled client ID is blocked by Google now.

**Comment by bergercookie** [comment]
> Thanks! I'm just having to create a new OAuth app, as the bundled client ID is blocked by Google now.

You're using the google tasks <> TW integration and it's blocked ? 
That 's not expected. What does the error message say?

**Comment by shymega** [comment]
@bergercookie Yes, so when I click to authorize the application with Google *Calendar*, not Tasks (long story), I get the following error:

> This app is blocked
> This app tried to access sensitive info in your Google Account. To keep your account safe, Google blocked this access.

I'm on mobile, so can't attach a screenshot from the browser at this time.

The client ID, I believe, is the one bundled with this repository.

The command I'm using is `tw_gcal_sync -f "pending" -c "Dom's Tasks"`.

Thanks.

**Comment by bergercookie** [comment]
Sorry for the very late reply.

> when I click to authorize the application with Google Calendar, not Tasks (long story), I get the following error:

This is expected.
This is because I haven't verified the Google Calendar <> TW integration with Google yet, only the Google Tasks <> TW.

**Comment by shymega** [comment]
Just realised my reply made it sound like I was using GTasks <=> TW. 

You're right. 

Do you think you'll be adding support for GCal <=> TW on the roadmap, or should I create my own OAuth app?

**Comment by bergercookie** [comment]
It's definitely in my TODO list, but not very high priority at the moment tbh. Adding your own OAuth app should hopefully be straightforward. If it isn't or if the instructions in this repo re this are off, please create a new ticket and I'll fix it! :)
